video_inference/gst/pipeline.py

```diff
@@ -0,0 +1,240 @@
+        """
+        Returns a string representation of the current pipeline.
+        This string is a valid pipeline and can be run with `gst-launch-1.0`.
+        """
+        self._format_pipeline()
+        pipeline_str = ""
+        for elem in self._pipeline:
+            pipeline_str += elem + " "
+            if elem == "!":
+                pipeline_str += "\\\n"
+        return pipeline_str
+
+    def _format_pipeline(self) -> None:
+        """
+        Updates pipeline by adding link operators between elements.
+        """
+        self._pipeline.clear()
+        self._pipeline.extend(self._elems[0])
+        for elem in self._elems[1:]:
+            if isinstance(elem, list):
+                self._pipeline.extend(["!", *elem])
+            else:
+                if elem != "t_data.":
+                    self._pipeline.append("!")
+                self._pipeline.append(elem)
+
+    def add_elements(self, *elements: str | list[str]) -> None:
+        self._elems.extend(elements)
+
+    def reset(self) -> None:
+        self._elems.clear()
+        self._pipeline.clear()
+
+    def run(
+        self,
+        run_prompt: str = "Running pipeline...",
+        print_err: bool = True,
+    ) -> bool:
+        """
+        Attempts to run current pipeline with `gst-launch-1.0` through a subprocess.
+
+        An erroneous pipeline will cause the subprocess to terminate with an exit message.
+
+        Pipeline can be shutdown with a SIGINT (KeyboardInterrupt) in which case a graceful exit is attempted.
+        The pipeline is forcefully shut down if the exit fails.
+
+        Returns:
+            bool: True if pipeline executed successfully, False if there was an error.
+        """
+        self._format_pipeline()
+        process = None
+        try:
+            if run_prompt:
+                print(run_prompt)
+            process = subprocess.Popen(
+                ["gst-launch-1.0", *self._pipeline],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                env=get_env(),
+            )
+            stdout, stderr = process.communicate()
+            if process.returncode != 0:
+                raise subprocess.CalledProcessError(
+                    process.returncode, process.args, output=stdout, stderr=stderr
+                )
+        except subprocess.CalledProcessError as e:
+            if print_err:
+                print(f"Pipeline failed with error: {e.stderr.decode()}")
+            return False
+        except KeyboardInterrupt:
+            print("\nShutting down pipeline...")
+            if process:
+                process.terminate()
+                try:
+                    process.wait(timeout=5)
+                except subprocess.TimeoutExpired:
+                    print("Shutdown failed, forcefully killing pipeline...")
+                    process.kill()
+                    process.wait()
+        return True
+
+
+class GstPipelineGenerator:
```

Comment:
Since multiple gstreamer pipeline include custom GStreamer plugins like synap-infer, synapoverlay etc
it would be helpful to provide a reference/readme to the GStreamer documentation(we should have it in astra?) for better context and understanding.